File: Deal/src/main/java/ru/leonov/deal/utility/ApplicationUtility.java
package ru.leonov.deal.utility;

import ru.leonov.deal.model.entity.ApplicationEntity;
import ru.leonov.deal.model.entity.ClientEntity;
import ru.leonov.deal.model.record.ApplicationHistoryElementRecord;

import java.time.LocalDate;
import java.util.ArrayList;

import static ru.leonov.deal.model.entity.ApplicationEntity.Status.PREAPPROVAL;

/**
 * Utility class that maintain {@link ApplicationEntity} needs.
 */
public class ApplicationUtility {

    private ApplicationUtility() {
        throw new IllegalStateException("Application service is utility class.");
    }

    /**
     * Update current status to {@link ApplicationEntity}. Also, append new status history item to application status
     * history.
     *
     * @param application application that needs to update status.
     * @param newStatus   new status of given application.
     */
    public static void updateApplicationStatus(ApplicationEntity application, ApplicationEntity.Status newStatus) {

        //updating current application status
        application.setStatus(newStatus);

        //adding new status to status history
        ApplicationHistoryElementRecord newHistoryElement = new ApplicationHistoryElementRecord();
        newHistoryElement.setDate(LocalDate.now());
        newHistoryElement.setStatus(newStatus);

        if (application.getStatusHistory() == null) {
            application.setStatusHistory(new ArrayList<>());
        }
        application.getStatusHistory().add(newHistoryElement);
    }

    /**
     * Create new application for given client.
     *
     * @param client client that requested new application.
     * @return new {@link ApplicationEntity} instance with PREAPPROVAL status and correct status history.
     */
    public static ApplicationEntity createNewApplication(ClientEntity client) {
        ApplicationEntity application = new ApplicationEntity();
        application.setClient(client);
        updateApplicationStatus(application, PREAPPROVAL);
        application.setCreationDate(LocalDate.now());

        return application;
    }
}

File: Deal/deal_impl/src/main/kotlin/ru/leonov/deal/utility/ApplicationUtility.kt
package ru.leonov.deal.utility

import ru.leonov.deal.model.entity.ApplicationEntity
import ru.leonov.deal.model.entity.ClientEntity
import ru.leonov.deal.model.enums.ApplicationStatusEnum
import ru.leonov.deal.model.record.ApplicationHistoryElementRecord
import java.time.LocalDate

/**
 * Update current status to [ApplicationEntity]. Also, append new status history item to application status
 * history.
 *
 * @param application application that needs to update status.
 * @param newStatus   new status of given application.
 */
fun updateApplicationStatus(application: ApplicationEntity, newStatus: ApplicationStatusEnum) {

    //updating current application status
    application.status = newStatus

    //adding new status to status history
    val newHistoryElement = ApplicationHistoryElementRecord(newStatus, LocalDate.now())
    application.statusHistory.add(newHistoryElement)
}

/**
 * Create new application for given client.
 *
 * @param client client that requested new application.
 * @return new [ApplicationEntity] instance with PREAPPROVAL status and correct status history.
 */
fun createNewApplication(client: ClientEntity): ApplicationEntity {
    return ApplicationEntity(
        client = client,
        status = ApplicationStatusEnum.PREAPPROVAL,
        creationDate = LocalDate.now(),
        statusHistory = mutableListOf(getNewStatus(ApplicationStatusEnum.PREAPPROVAL))
    )
}

/**
 * Generate new [ApplicationHistoryElementRecord] with new status and current date.
 */
fun getNewStatus(newStatus: ApplicationStatusEnum): ApplicationHistoryElementRecord {
    return ApplicationHistoryElementRecord(newStatus, LocalDate.now())
}
